staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,569 @@
+// should try to find a different QueueSet to use; execute and
+// afterExecution are irrelevant in this case.  Otherwise, if execute
+// then the client should start executing the request and, once the
+// request finishes execution or is canceled, call afterExecution().
+// Otherwise the client should not execute the
+// request and afterExecution is irrelevant.
+func (qs *queueSet) Wait(ctx context.Context, hashValue uint64) (tryAnother, execute bool, afterExecution func()) {
+	var req *fq.Request
+	shouldReturn, tryAnother, execute, afterExecution := func() (
+		shouldReturn, tryAnother, execute bool, afterExecution func()) {
+
+		qs.lockAndSyncTime()
+		defer qs.lock.Unlock()
+		// A call to Wait while the system is quiescing will be rebuffed by
+		// returning `tryAnother=true`.
+		if qs.emptyHandler != nil {
+			return true, true, false, nil
+		}
+
+		// ========================================================================
+		// Step 1:
+		// 1) Start with shuffle sharding, to pick a queue.
+		// 2) Reject old requests that have been waiting too long
+		// 3) Reject current request if there is not enough concurrency shares and
+		// we are at max queue length
+		// 4) If not rejected, create a request and enqueue
+		req = qs.timeoutOldRequestsAndRejectOrEnqueueLocked(hashValue)
+		// req == nil means that the request was rejected - no remaining
+		// concurrency shares and at max queue length already
+		if req == nil {
+			metrics.AddReject(qs.config.Name, "queue-full")
+			return true, false, false, func() {}
+		}
+
+		// ========================================================================
+		// Step 2:
+		// 1) The next step is to invoke the method that dequeues as much as possible.
+
+		// This method runs a loop, as long as there
+		// are non-empty queues and the number currently executing is less than the
+		// assured concurrency value.  The body of the loop uses the fair queuing
+		// technique to pick a queue, dequeue the request at the head of that
+		// queue, increment the count of the number executing, and send true to
+		// the request's channel.
+		qs.dequeueWithChannelAsMuchAsPossibleLocked()
+		return false, false, false, func() {}
+	}()
+	if shouldReturn {
+		return tryAnother, execute, afterExecution
+	}
+
+	// ========================================================================
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud request's channel
+	// and return appropriately.  While waiting this thread does no additional
+	// work so we decrement the go routine counter
+	qs.counter.Add(-1)
+
+	select {
+	case execute := <-req.DequeueChannel:
+		if execute {
+			// execute the request
+			return false, true, func() {
+				qs.finishRequestAndDequeueWithChannelAsMuchAsPossible(req)
+			}
+		}
+		klog.V(5).Infof("request timed out after being enqueued\n")
+		metrics.AddReject(qs.config.Name, "time-out")
```

Comment:
Why don't we need to decrement the counter here? Wouldn't it have been incremented prior to sending something down the dequeue channel?